### ISSUE TITLE
woof-next: bootstrapped packages can't be configured

### ISSUE THREAD

**Issue by s243a** [puppylinux-woof-CE/woof-CE#1387]
Apologies for posting this here but I can't see how to open an issue on jamesbond's woof-next fork. 

Anyway, I'm guessing that part of the reason for the speed of woof-next is that all configuration is done after all the packages have been extracted, with the exception that:
1. packages installed from the "rootfs-packages" directory are configured when they are installed via the pinstall.sh script,
2. packages that are installed via bootstrap are never configured, since the bootstrap function doesn't copy the post install script anywhere:

```
bootstrap_install() {
	local data decompressor
	if [ -z "$WITHOUT_DPKG" ]; then
		dpkg-deb -X "$REPO_DIR/$PKGFILE" $CHROOT_DIR
	else
		data=$(ar t "$REPO_DIR/$PKGFILE" | grep data)
		case $data in
			*xz) decompressor="unxz -c" ;;
			*gz) decompressor="gunzip -c" ;;
			*bz2) decompressor="bunzip2 -c" ;;
			*lzma) decompressor="unlzma -c" ;;
		esac
		ar p "$REPO_DIR/$PKGFILE" $data | $decompressor | tar -xv -C $CHROOT_DIR
	fi |
	sed '1 s|^.*$|/.|; s|^\.||' > "$CHROOT_DIR$ADMIN_DIR/info/${PKG}.list" &&
	update_pkg_status "$PKG" "$PKGPRIO" "$PKGSECTION" "$PKGVER" "$PKGDEP"
}
```
https://github.com/puppylinux-woof-CE/woof-CE/blob/9e35bf138c2601f203ca5ea73182901fc5704625/builders/deb-build.sh#L269 

Regarding 1. We can call:
`%dpkg_configure --force-all -a `
https://github.com/puppylinux-woof-CE/woof-CE/blob/9e35bf138c2601f203ca5ea73182901fc5704625/woof-distro/x86/devuan/ascii/basesfs#L149

always prior to calling %addpkg so that the packages get configured in a logical order. Maybe a note about this should be added to the readme. 

Regarding 2. we could call the %reinstall directive (as a way to configure the package after the fact) but this would negate the speed advantage of the %bootstrap directive.

**Comment by s243a** [comment]
To avoid creating too many "woof-next" issues, I'll post multiple issues in this thread unless someone suggests breaking them up into separate thread. 

New issue: **cutdown script causes exit of deb-build.sh if  locale directory either doesn't exist or is empty.** 

Here is some output "with sh -x on"
```

+ '[' zu '!=' en ']'
+ mv chroot-ascii-i386/usr/share/locale/zu nls-holder/usr/share/locale
++ ls chroot-ascii-i386/usr/lib/locale
ls: cannot access 'chroot-ascii-i386/usr/lib/locale': No such file or directory
+ read -r p
+ return 0
+ cleanup
+ rm -f /tmp/tracker.1090 /tmp/flatten.1090
+ exit
```
[/builders/deb-build.sh#L220](https://github.com/puppylinux-woof-CE/woof-CE/blob/9e35bf138c2601f203ca5ea73182901fc5704625/builders/deb-build.sh#L220)

It wouldn't be too hard to check if the directory is empty or non-existant and either continue or output a better error message.  Anyway, the exit is probably due to the trap statment:

```
trap 'cleanup; exit' INT HUP TERM 0
cleanup() {
	rm -f $TRACKER $FLATTEN
}
```
[/builders/deb-build.sh#L49](https://github.com/puppylinux-woof-CE/woof-CE/blob/9e35bf138c2601f203ca5ea73182901fc5704625/builders/deb-build.sh#L49)

**Comment by jamesbond3142** [comment]
> Apologies for posting this here but I can't see how to open an issue on jamesbond's woof-next fork.

No worries. This is the correct place to report issues about woof-next.

> 1. packages installed from the "rootfs-packages" directory are configured when they are installed via the pinstall.sh script

Yes, because this is our __own__ packages and we are supposed to know how to configure it (i.e. write a proper post-install script that understand the environment it is running in, and will not attempt to destroy the host's system).

> 2. packages that are installed via bootstrap are never configured, since the bootstrap function doesn't copy the post install script anywhere.

Yes. The reasons that underlies this decision are:
1. Most packages don't have install scripts.
2. Some package post install scripts require circular dependency (an hypothetical example: let assume libc6's post-install requires coreutils, and coreutils post-install uses binaries that depends on libc6).

So its better to install first and dpkg-reconfigure later. In fact, that's how debootstrap actually does it.

> Regarding 1. We can call: %dpkg_configure --force-all -a always prior to calling %addpkg so that the packages get configured in a logical order. Maybe a note about this should be added to the readme.

This may or may not be necessary, so it's not exactly a requirement. 

> Regarding 2. we could call the %reinstall directive (as a way to configure the package after the fact) but this would negate the speed advantage of the %bootstrap directive.

Or, once "dpkg" has been installed in the chroot, just use %dpkgchroot to enable it. Subsequent installation will use that dpkg-in-chroot to install stuff and run the post-install scripts properly. You get the package installed right the first time round, no need for %reinstall.

This will of course slow down the builds, and the reason why I stick with %bootstrap most of the time is because my observation that most packages don't have any post-install or configuration phase; but if you want to build everything "properly", then that's how you can do it. That's how debootstrap does it: once a "minimal" stuff is installed, all next package installation will use dpkg.

> New issue: cutdown script causes exit of deb-build.sh if locale directory is either doesn't exist or is empty.

Thanks, let me have a look.

**Comment by s243a** [comment]
> > 1. packages installed from the "rootfs-packages" directory are configured when they are installed via the pinstall.sh script
> 
> Yes, because this is our **own** packages and we are supposed to know how to configure it (i.e. write a proper post-install script that understand the environment it is running in, and will not attempt to destroy the host's system).

This presumes the person using woof-next understands this. I put the install script for sc0tmann's package manager (i.e. puppy-PKG) in there and I discovered that it installed stuff to my host system. Worse yet is when people who don't understand how this folder is used share their build system. 

Anyway, I think this is risky. I took a different philosophy here:
I posted this on the puppy forum:





> Now regarding the pinstall script in /woof-CE/tree/woof-next/woof-code/rootfs-packages
> 
> a given post install script might not be robust with respect to where it is located and what kind of environment it is executed in. I defined two modes which are:
> `PINSTALL_MODE='chroot' #Mean the script is meant to be run in a chroot enviornment`
>  
> [/builders/deb-build.sh#L26](https://github.com/s243a/woof-CE/blob/f564fa4214feedf69329d8d51c08f89ba41aec8c/builders/deb-build.sh#L26)
> and also
>  `$PINSTALL_MODE = 'relative' #Which means that the rootfs is relative to the current director and the script should not be executed in a chroot environment.`
> [/builders/deb-build.sh#L347](https://github.com/s243a/woof-CE/blob/f564fa4214feedf69329d8d51c08f89ba41aec8c/builders/deb-build.sh#L347)
> 
> To use 'relative' mode one must set SAFE_MODE=0 (SAFE_MODE=1 is the default).
> Code:	
> 
```
> #When SAFE_MODE=1, the pinstall script (in a directory based install) cannot change
> #the mode between relative vs chroot
> #s243a:TODO add directive to turn safe mode on and off. Give prompt to indicate that
> #      safe mode is being turned off.
> export SAFE_MODE=1 #Not sure if we need to export this
```
> [/builders/deb-build.sh#L15](https://github.com/s243a/woof-CE/blob/f564fa4214feedf69329d8d51c08f89ba41aec8c/builders/deb-build.sh#L15)
> 
> I plan to add directives to change SAFE_MODE and the PINSTALL_MODE. For security reasons I'll add a prompt so that someone can verify that they want to change the safe mode.
[ alternative puppy build system thread by wander,p=1028494#1028494 ](http://murga-linux.com/puppy/viewtopic.php?p=1028494#1028494) 

 
> 
> > 1. packages that are installed via bootstrap are never configured, since the bootstrap function doesn't copy the post install script anywhere.
> 
> Yes. The reasons that underlies this decision are:
> 
>     1. Most packages don't have install scripts.
> 
>     2. Some package post install scripts require circular dependency (an hypothetical example: let assume libc6's post-install requires coreutils, and coreutils post-install uses binaries that depends on libc6).
I think that in some cases you can build a complete system and have it mostly work without running the post install scripts. I think mistfire's early versions of TazPup did this but her more recent builds utilize the package manger which calls the post install scirpts. 

So I don't necessarily disagree here, but I think anyone working on woof-next should keep this in mind for when more extensive documentation might be written for woof-next. 
> 
> So its better to install first and dpkg-reconfigure later. In fact, that's how debootstrap actually does it.
> 
> > Regarding 1. We can call: %dpkg_configure --force-all -a always prior to calling %addpkg so that the packages get configured in a logical order. Maybe a note about this should be added to the readme.
> 
> This may or may not be necessary, so it's not exactly a requirement.
> 
> > Regarding 2. we could call the %reinstall directive (as a way to configure the package after the fact) but this would negate the speed advantage of the %bootstrap directive.
> 
> 
> > New issue: cutdown script causes exit of deb-build.sh if locale directory is either doesn't exist or is empty.
> 
> Thanks, let me have a look.

Here is my proposal:
```
nls)
  if [ -d "$CHROOT_DIR/usr/share/locale" ] &&
	 [ `ls | grep -c -m 1 '.*'` -eq 1 ]; then 
	[ -d $CHROOT_DIR/usr/lib64/locale ] && LIBDIR=lib64
	rm -rf $NLS_DIR; mkdir -p $NLS_DIR/usr/share/locale $NLS_DIR/usr/$LIBDIR/locale
	for p in $(ls $CHROOT_DIR/usr/share/locale); do
		[ $p != en ] && mv $CHROOT_DIR/usr/share/locale/$p $NLS_DIR/usr/share/locale
	done
  fi
  if [ -d "$CHROOT_DIR/usr/$LIBDIR/locale" ] &&
	 [ `ls | grep -c -m 1 '.*'` -eq 1 ]; then 
	for p in $(ls $CHROOT_DIR/usr/$LIBDIR/locale); do #TODO: add check to see if the directory is empty, otherwise the script might unexpectly exit
		case $p in
			en_US|en_AU|en_US.*|en_AU.*|C|C.*) ;; # skip
			*) mv $CHROOT_DIR/usr/$LIBDIR/locale/$p $NLS_DIR/usr/$LIBDIR/locale
		esac
	done 
  fi ;;
```
[/builders/deb-build.sh#L246 (s243a's version)](https://github.com/s243a/woof-CE/blob/6df408b884783e705f95ba16156b428e7d19c85d/builders/deb-build.sh#L246 )
I'm testing it now.

**Comment by jamesbond3142** [comment]
> This presumes the person using woof-next understands this. I put the install script for sc0tmann's package manager (i.e. puppy-PKG) in there and I discovered that it installed stuff to my host system. Worse yet is when people who don't understand how this folder is used share their build system.

Yes, this isn't made very clear, because I operated on the assumption that people using the build system would at least be familiar with the inner workings of PET package manager, since rootfs-packages contains exploded version of PETs.

In a PET, the pinstall.sh __should__ operate assuming that the current directory is the system root. This has been the convention, but many PETs don't do this properly - if you look at commit histories of Woof-CE, there are many of such packages, unfortunately.

The proper fix is to notify Scott to make sure that his Pkg PET's pinstall.sh work correctly.

> Anyway, I think this is risky. I took a different philosophy here:

Yes, I saw that. I wouldn't dismiss it, but I would note that running "chroot" would only work if the chroot directory already have enough "meats" in order to run the pinstall script (e.g. bash, libc, some common tools like sed, awk. cut, etc) otherwise it won't work.

> Here is my proposal:

I will check. I don't have time to look into woof-next yet, but I'll keep this in mind when I do. Thanks again.

**Comment by jamesbond3142** [comment]
>> Here is my proposal:

> I will check. I don't have time to look into woof-next yet, but I'll keep this in mind when I do. Thanks again.

I fix it in a different way. See if it works for you. Cheers.

**Comment by s243a** [comment]
> > > Here is my proposal:
> 
> > I will check. I don't have time to look into woof-next yet, but I'll keep this in mind when I do. Thanks again.
> 
> I fix it in a different way. See if it works for you. Cheers.

I'll check it out :)

On a possibly related note to the script exiting early, I'm wondering if there needs to be a blank line at the end of the basesfs file. After my fix I noticed that the last line
`%makesfs iso/iso-root/puppy.sfs -comp gzip # -Xcompression-level 1`
wasn't getting read, in the flatten function. 

I added a black line at the end of this file and I'm going to see what happens.

**Comment by jamesbond3142** [comment]
> On a possibly related note to the script exiting early, I'm wondering if there needs to be a blank line at the end of the basesfs file. After my fix I noticed that the last line
> `%makesfs iso/iso-root/puppy.sfs -comp gzip # -Xcompression-level 1`
> wasn't getting read, in the flatten function.
> 
> I added a blank line at the end of this file and I'm going to see what happens.
> 
> **Edit:** Adding the blank line achieved the intended result :) That said I don't understand why this should be necessary.

I'm not sure what do you mean by blank line. The screenshot shows the default basesfs, as edited in geany. This basesfs does get processed properly as expected.

![xscreenshot-20190603T223848](https://user-images.githubusercontent.com/2094782/58802472-9c578d80-8650-11e9-9714-dc8df48ba4af.png)

You may notice that line 162 is empty (blank). But this is always the case with geany. You cannot have a line standing by itself. Try to create a one-line text file in document. Notice that you always have two lines - the first line, and the blank second line. Are you using geany or are you using other editors to create the file? 

Anyway, the reason for that "extra blank line" (in geany) is because each line always ends with a "newline" character. That's why geany shows it. That newline character is also the character that tells the shell's "read" function to finish reading the line. A line that does not end with the "newline" character isn't going to be read by the the shell.
